src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/FSharp/FSharpOptionConverter.cs

```diff
@@ -0,0 +1,57 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Text.Json.Serialization.Metadata;
+
+namespace System.Text.Json.Serialization.Converters
+{
+    // Converter for F# optional values: https://fsharp.github.io/fsharp-core-docs/reference/fsharp-core-option-1.html
+    // Serializes `Some(value)` using the format of `value` and `None` values as `null`.
+    internal sealed class FSharpOptionConverter<TOption, TElement> : JsonResumableConverter<TOption>
+        where TOption : class
+    {
+        // While technically not implementing IEnumerable, F# optionals are effectively generic collections of at most one element.
+        internal override ConverterStrategy ConverterStrategy => ConverterStrategy.Enumerable;
+        internal override Type? ElementType => typeof(TElement);
+
+        private readonly Func<TOption, TElement> _optionValueGetter;
+        private readonly Func<TElement?, TOption> _optionConstructor;
+
+        [RequiresUnreferencedCode(FSharpCoreReflectionProxy.FSharpCoreUnreferencedCodeMessage)]
+        public FSharpOptionConverter(JsonConverter<TElement> elementConverter)
+        {
+            _optionValueGetter = FSharpCoreReflectionProxy.Instance.CreateFSharpOptionValueGetter<TOption, TElement>();
+            _optionConstructor = FSharpCoreReflectionProxy.Instance.CreateFSharpOptionSomeConstructor<TOption, TElement>();
+            // If the element converter is value, this converter will also be writing values
+            // Set a flag to signal this fact to the converter infrastructure.
+            CanWriteJsonValues = elementConverter.ConverterStrategy == ConverterStrategy.Value;
+        }
+
+        internal override bool OnTryRead(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options, ref ReadStack state, out TOption? value)
+        {
+            state.Current.JsonPropertyInfo = state.Current.JsonTypeInfo.ElementTypeInfo!.PropertyInfoForTypeInfo;
+            var elementConverter = (JsonConverter<TElement>)state.Current.JsonPropertyInfo.ConverterBase;
+
+            if (elementConverter.TryRead(ref reader, typeof(TElement), options, ref state, out TElement? element))
```

Comment:
If `false` is returned (for the `Stream` cases), how does this resume? It seems like a state variable is necessary.

Is the `Stream` case tested?

Reply:
It is, cf. https://github.com/dotnet/runtime/pull/55108/files/2e793422dca84bd22d55cdfa2cd6c9b6c5d4963e#diff-9fb2d62d8f334fd5a7bba90334713825cbcd440eca0177c4d387c93055c74f23R134-R154

I don't believe we need to store any state here, the converter only needs to pass the fully deserialized element to the optional constructor once it has completed.